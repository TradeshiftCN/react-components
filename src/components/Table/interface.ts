export type Key = string | number;

export type Order = 'desc' | 'asc';

export interface Column<T> {
	key?: Key;
	dataIndex?: string;
	width?: string | number;
	fixed?: 'left' | 'right' | boolean;
	colSpan?: number;
	className?: string;
	title: React.ReactNode | string;
	align?: string;
	searchDebounceTime?: number;
	/** 响应 search 的操作，设为 true 则使用默认配置 */
	searchTriggers?: Array<'blur' | 'input' | 'enter'> | boolean;
	/** 排序的方法，设为 true 时仅显示排序按钮 */
	sorter?: ((a: T, b: T) => number) | boolean;
	sortOrder?: Order | false;
	sortDirections?: Array<Order>;
	defaultSortOrder?: Order;
	onCell?(record: T, index: number): any;
	onHeaderCell?(column: Column<T>): any;
	render?(value: any, row: T, index: number): React.ReactNode | string;
}

export interface RowSelection<T> {
	onChange(keys: Array<Key>, rows: Array<T>): void;
	selectedRowKeys: Array<Key>;
	type?: 'checkbox';
}

export interface ExpandIconProps<T> {
	prefixCls: string;
	expanded: boolean;
	record: T;
	needIndentSpaced: boolean;
	expandable: boolean;
	onExpand: (record: T, event?: React.MouseEvent) => void;
}
