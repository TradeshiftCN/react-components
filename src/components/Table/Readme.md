### Basic Usage

```jsx
const columns = [
	{
		title: 'One',
		dataIndex: 'one'
	},
	{
		title: 'Align right',
		dataIndex: 'two',
		align: 'right'
	},
	{
		title: 'CUstom th/td props',
		dataIndex: 'three',
		onCell(row) {
			return {
				style: {
					color: 'red'
				},
				title: `id: ${row.id}`
			};
		},
		onHeaderCell(column) {
			return {
				style: {
					color: 'green'
				},
				title: `id: ${column.dataIndex}`
			};
		}
	},
	{
		title: 'Action',
		key: 'action',
		render() {
			return (
				<div>
					<a>edit</a> | <a>delete</a>
				</div>
			);
		}
	}
];

const data = [
	{
		id: 1,
		one: 'A',
		two: 'B',
		three: 'C'
	},
	{
		id: 2,
		one: 'D',
		two: 'E',
		three: 'F'
	},
	{
		id: 3,
		one: 'G',
		two: 'H',
		three: 'I'
	}
];

<Table columns={columns} data={data} rowKey="id" />;
```

### Selection

```jsx
const columns = [
	{
		title: 'One',
		dataIndex: 'one'
	},
	{
		title: 'Two',
		dataIndex: 'two'
	},
	{
		title: 'Three',
		dataIndex: 'three'
	}
];

const data = [
	{
		id: 1,
		one: 'A',
		two: 'B',
		three: 'C'
	},
	{
		id: 2,
		one: 'D',
		two: 'E',
		three: 'F'
	},
	{
		id: 3,
		one: 'G',
		two: 'H',
		three: 'I'
	}
];

initialState = { selectedRowKeys: [] };

<Table
	columns={columns}
	data={data}
	rowKey="id"
	rowSelection={{
		selectedRowKeys: state.selectedRowKeys,
		onChange: keys => setState({ selectedRowKeys: keys })
	}}
/>;
```

### Search

```jsx
const columns = [
	{
		title: 'Front-end search',
		dataIndex: 'one',
		searchTriggers: true,
		width: 200
	},
	{
		title: 'Back-end search',
		dataIndex: 'two',
		searchDebounceTime: 500,
		searchTriggers: ['enter', 'input']
	},
	{
		title: 'No search',
		dataIndex: 'three'
	}
];

const data = [
	{
		id: 1,
		one: 'press enter to search',
		two: 'search when stop input',
		three: 'C'
	},
	{
		id: 2,
		one: 'no debounce',
		two: 'delay 500ms',
		three: 'F'
	},
	{
		id: 3,
		one: 'G',
		two: 'H',
		three: 'I'
	}
];

const onChange = query => {
	setState({
		data: data.filter(v =>
			Object.keys(query.search).every(key => v[key].includes(query.search[key]))
		)
	});
};

initialState = { data: data };

<Table columns={columns} data={state.data} rowKey="id" onChange={onChange} />;
```

### Sort

```jsx
const columns = [
	{
		title: 'Local sort',
		dataIndex: 'one',
		sorter: (a, b) => a.one - b.one,
		sortDirections: ['desc', 'asc'] // 默认值 - 降序/升序/不排序
	},
	{
		title: 'Server sort',
		dataIndex: 'two',
		sorter: true,
		sortDirections: ['asc'] // 升序/不排序
	}
];
const data = [
	{
		id: 1,
		one: 1,
		two: 3
	},
	{
		id: 2,
		one: 2,
		two: 2
	},
	{
		id: 3,
		one: 3,
		two: 1
	}
];
initialState = { data: data };

<Table
	columns={columns}
	data={state.data}
	rowKey="id"
	onChange={({ sort }) => {
		const { field, order } = sort;

		// 这里只 mock 第二列
		if (field === 'two') {
			setTimeout(() => {
				setState({
					data: order === 'asc' ? data.slice().reverse() : data
				});
			}, 500);
		}
	}}
/>;
```

### Fixed Header

```jsx
const data = [];
const columns = [
	{
		title: 'name',
		dataIndex: 'name',
		width: 1500
	},
	{
		title: 'id',
		dataIndex: 'key',
		width: 100
	}
];
for (let i = 1; i <= 100; i++) {
	data.push({
		name: `row ${i} `.repeat(49),
		key: i
	});
}
<Table
	data={data}
	columns={columns}
	rowKey="key"
	scroll={{
		y: 260,
		x: 1600
	}}
/>;
```

### Expandable Row

_Note: If you want to hide expand icon of special row, you can add a className `ts-rc-table-row--unexpandable` [Set Table expandable property at row level](https://github.com/ant-design/ant-design/issues/7817)_

```jsx
const data = [
	{
		id: 1,
		data: [11, 12, 13]
	},
	{
		id: 2,
		data: [21]
	},
	{
		id: 3
	}
];
const columns = [
	{
		title: 'id',
		dataIndex: 'id'
	}
];
const renderSubRow = row => {
	if (row.data) {
		return (
			<table>
				<tbody>
					{row.data.map(v => (
						<tr key={v}>
							<td>child: {v}</td>
						</tr>
					))}
				</tbody>
			</table>
		);
	}
};
<Table
	data={data}
	columns={columns}
	rowKey="id"
	rowClassName={row => {
		if (!row.data) {
			return 'ts-rc-table-row--unexpandable';
		}
	}}
	expandedRowRender={renderSubRow}
	defaultExpandedRowKeys={[1, 2, 3]}
/>;
```

P.S. 如果列配置中同时启用了 sorter 和 searchTriggers，仅有**查询**功能会生效（因为 ts-ui 的布局冲突）

### Interface

```ts static
export type Order = 'desc' | 'asc';

export interface Column<T> {
	key?: Key;
	dataIndex?: string;
	width?: string | number;
	fixed?: 'left' | 'right' | boolean;
	colSpan?: number;
	className?: string;
	title: React.ReactNode | string;
	align?: string;
	searchDebounceTime?: number;
	/** 响应 search 的操作，设为 true 则使用默认配置 */
	searchTriggers?: Array<'blur' | 'input' | 'enter'> | boolean;
	/** 排序的方法，设为 true 时仅显示排序按钮 */
	sorter?: ((a: T, b: T) => number) | boolean;
	sortOrder?: Order | false;
	sortDirections?: Array<Order>;
	defaultSortOrder?: Order;
	onCell?(record: T, index: number): any;
	onHeaderCell?(column: Column<T>): any;
	render?(value: any, row: T, index: number): React.ReactNode | string;
}
```
