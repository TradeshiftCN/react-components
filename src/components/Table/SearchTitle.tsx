import React, { Component } from 'react';
import cx from 'classnames';
import _ from 'lodash';
import { Column, Key } from './interface';

type SearchTitleProps<T> = {
	column: Column<T>;
	prefixCls: string;
	activeColumn?: Key;
	isActive: boolean;
	columnKey: Key;
	setActiveColumn(key?: Key): void;
	value: string;
	/** 用于和上层传入的 value 双向绑定 */
	onChange(text: string): void;
	onSearch(text: string): void;
};

type SearchTitleState = {
	originTitleWidth: number;
	debounceSearch?(text: string): void;
};

type EventCallback = (e: Event) => void;

let activeSearchTitle: SearchTitle<any>;

document.addEventListener('click', (e: MouseEvent) => {
	const el = e.target;
	if (
		el &&
		activeSearchTitle &&
		activeSearchTitle.rootRef &&
		!activeSearchTitle.rootRef.contains(el as Node)
	) {
		activeSearchTitle.props.setActiveColumn();
	}
});

class SearchTitle<T> extends Component<SearchTitleProps<T>, SearchTitleState> {
	static defaultTriggers = ['enter'];
	private titleTextRef?: HTMLSpanElement | null;
	private inputRef?: HTMLInputElement | null;
	public rootRef?: HTMLDivElement | null;

	state: SearchTitleState = {
		originTitleWidth: 80
	};

	render() {
		const { column, prefixCls, setActiveColumn, isActive, columnKey, value } = this.props;
		const events = this.generateEventProps();
		const showInput = isActive || !!value;
		const activeEvents: { [key: string]: () => void } = {};

		if (!showInput) {
			activeEvents.onClick = () => {
				activeSearchTitle = this;
				setActiveColumn(columnKey);
			};
		}

		return (
			<div
				ref={ref => (this.rootRef = ref)}
				className={cx({
					[`${prefixCls}-search-title`]: true,
					[`${prefixCls}-search-title--input`]: showInput
				})}
				{...activeEvents}
			>
				<i className={cx('ts-icon-search', `${prefixCls}-search-title-search-icon`)}></i>
				<span ref={ref => (this.titleTextRef = ref)}>
					{showInput ? (
						<input
							style={{ width: this.state.originTitleWidth }}
							ref={ref => (this.inputRef = ref)}
							value={value}
							autoFocus
							{...events}
						/>
					) : (
						column.title
					)}
				</span>
				{!!value && (
					<i
						className={cx('ts-icon-remove', `${prefixCls}-search-title-clear-icon`)}
						onClick={() => {
							this.props.onChange('');
							(this.state.debounceSearch || this.search)('');
							setActiveColumn();
						}}
					></i>
				)}
			</div>
		);
	}

	componentDidMount() {
		this.updateOriginTitleWidth();
		this.updateDebounceSearch();
	}

	componentDidUpdate(prevProps: SearchTitleProps<T>) {
		if (prevProps.isActive && !this.props.isActive) {
			this.updateOriginTitleWidth();
		}

		if (prevProps.column.searchDebounceTime !== this.props.column.searchDebounceTime) {
			this.updateDebounceSearch();
		}
	}

	private updateOriginTitleWidth() {
		const titleText = this.titleTextRef;
		const columnWidth = this.props.column.width;
		let width = 80;

		if (columnWidth && _.isNumber(columnWidth)) {
			width = (this.props.column.width as number) - 20 - 30; // parent padding and search btn
		} else if (titleText && !this.props.isActive) {
			width = titleText.getBoundingClientRect().width;
		}

		width -= 30; // clear btn

		if (width > 0) {
			this.setState({
				originTitleWidth: width
			});
		}
	}

	private search = (text: string) => {
		this.props.onSearch(text);
	};

	private updateDebounceSearch() {
		const debounceTime = this.props.column.searchDebounceTime;
		this.setState({
			debounceSearch: debounceTime ? _.debounce(this.search, debounceTime) : undefined
		});
	}

	private generateEventProps() {
		const events: { [key: string]: EventCallback } = {};
		const { column } = this.props;
		const callback = this.state.debounceSearch || this.search;
		const updateValue = this.props.onChange;
		const triggers =
			column.searchTriggers === true ? SearchTitle.defaultTriggers : column.searchTriggers;

		(triggers as string[]).forEach(trigger => {
			let k: string,
				v: EventCallback = (e: Event) => callback((e.target as HTMLInputElement).value);
			switch (trigger) {
				case 'input':
					k = 'onChange';
					v = (e: Event) => {
						const text = (e.target as HTMLInputElement).value;
						updateValue(text);
						callback(text);
					};
					break;
				case 'blur':
					k = 'onBlur';
					break;
				case 'enter':
					k = 'onKeyDown';
					v = (e: Event) => {
						if ((e as KeyboardEvent).keyCode === 13) {
							const text = (e.target as HTMLInputElement).value;
							callback(text);
						}
					};
					break;
				default:
					return;
			}
			events[k] = v;
		});

		if (!events.onChange) {
			events.onChange = (e: Event) => updateValue((e.target as HTMLInputElement).value);
		}

		return events;
	}
}

export default SearchTitle;
