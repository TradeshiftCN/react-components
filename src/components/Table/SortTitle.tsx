import React from 'react';
import { Column } from './interface';
import cx from 'classnames';

type SortTitleProps<T> = { column: Column<T>; prefixCls: string; order: 'asc' | 'desc' | false };

function SortTitle<T>(props: SortTitleProps<T>) {
	const { prefixCls, column, order } = props;

	return (
		<div className={`${prefixCls}-sort-title`}>
			{column.title}
			{order && (
				<i
					className={cx({
						[`${prefixCls}-sort-title-icon`]: true,
						[`${prefixCls}-sort-title-icon--${order}`]: !!order,
						'ts-icon-triangledown': order === 'desc',
						'ts-icon-triangleup': order === 'asc'
					})}
				></i>
			)}
		</div>
	);
}

export default SortTitle;
