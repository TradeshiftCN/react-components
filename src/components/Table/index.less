@import '../../style/variables/index';

.ts-rc-table {
	position: relative;
	font-size: @ts-fontsize;
	background-color: @ts-color-white;
	table {
		width: 100%;
		line-height: @ts-unit;
		border-collapse: collapse;
		border-spacing: 0;
	}

	th,
	td {
		vertical-align: inherit;
	}

	&-layout-fixed table {
		table-layout: fixed;
	}

	&-thead {
		font-size: @ts-fontsize-small;
		text-transform: uppercase;
		background-color: @ts-color-white;
		& > tr > th {
			padding: @ts-unit-half @ts-unit-half @ts-unit-half - 1px;
			border-bottom: 1px solid @ts-color-gray-light;
			text-align: left;
			font-weight: @ts-fontweight-semibold;
			&.ts-rc-table-sort-cell {
				padding-right: @ts-unit-double;
			}
		}
	}

	&-tbody {
		background-color: @ts-color-white;
		& > tr > td {
			padding: @ts-unit-half @ts-unit-half @ts-unit-half - 1px;
			border-bottom: 1px solid @ts-color-gray-lighter;
		}

		& > tr:last-child > td {
			border-bottom-color: transparent;
		}
	}

	&-row-selected {
		background-color: @ts-color-blue-lightest;
	}

	&-selection-column {
		cursor: pointer;
	}

	&-selection-icon {
		color: @ts-color-gray-light;
		font-size: 27px;
		vertical-align: middle;
		&--checked {
			color: @ts-color-blue;
		}
	}

	th.ts-rc-table-search-cell {
		position: relative;
	}

	&-search-title {
		padding-left: 1.5 * @ts-unit;
		&--input {
			padding-right: 1.5 * @ts-unit;
			font-weight: normal;
		}
		input {
			display: inline-block;
			width: auto;
			height: @ts-unit;
			border: 0;
		}
		i {
			position: absolute;
			display: inline-block;
			width: 2 * @ts-unit;
			height: 2 * @ts-unit;
			line-height: 2 * @ts-unit;
		}
		&-search-icon {
			left: 0;
			top: 0;
		}
		&-clear-icon {
			right: 0;
			top: 0;
			color: @ts-color-slate-lightest;
			cursor: pointer;
		}
	}

	th.ts-rc-table-sort-cell {
		cursor: pointer;
		position: relative;
	}

	&-sort-cell {
		&--active {
			background-color: @ts-color-gray-lightest;
		}
	}

	&-sort-title {
		user-select: none;
		&-icon {
			display: inline-block;
			width: 2 * @ts-unit;
			height: 2 * @ts-unit;
			line-height: 2 * @ts-unit;
			font-size: 14px;
			position: absolute;
			right: 0;
			top: 0;
		}
	}

	&-placeholder {
		text-align: center;
	}
}
