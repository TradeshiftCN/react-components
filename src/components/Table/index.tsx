import React, { Component } from 'react';
import cx from 'classnames';
import RcTable from 'rc-table';
import _ from 'lodash';
import { Column, RowSelection, Order } from './interface';
import SearchTitle from './SearchTitle';
import SortTitle from './SortTitle';

type TableProps<T> = {
	/** 数据 */
	data: Array<T>;
	/** 列配置 */
	columns: Array<Column<T>>;
	/** 表格行的唯一 id 的 key */
	rowKey: keyof T;
	/** 修改前缀会丢失默认样式 */
	prefixCls?: string;
	rowSelection?: RowSelection<T>;
	/** 自定义行 className */
	rowClassName?: ((record: T, index: number, indent: string) => string) | string;
	/** 无数据时显示 */
	emptyText?: string | React.ReactNode;
	className?: string;
	/** identifier of the container div */
	id?: string;
	onChange?(query: { search: SearchData; sort: SortData<T> }): void;
};

type TableState<T> = {
	searchActiveColumn?: string | number;
	searchData: SearchData;
	sortData: SortData<T>;
};

type SortData<T> = {
	field?: string | number;
	order?: Order | false;
	sorter?: Column<T>['sorter'];
};
type SearchData = { [key: string]: string; [key: number]: string };

const checkboxClassName = (prefixCls: string, isSelected: boolean) =>
	cx({
		'ts-icon-checkbox': !isSelected,
		'ts-icon-checkboxon': isSelected,
		[`${prefixCls}-selection-icon`]: true,
		[`${prefixCls}-selection-icon--checked`]: isSelected
	});

class Table<T> extends Component<TableProps<T>, TableState<T>> {
	static defaultProps = {
		prefixCls: 'ts-rc-table',
		emptyText: ''
	};
	static defaultSortDirections: Order[] = ['asc', 'desc'];

	state: TableState<T> = {
		searchData: {},
		searchActiveColumn: undefined,
		sortData: {}
	};

	render() {
		const props = this.props;
		const columns = this.renderRowSelections(this.renderSearch(this.renderSort(props.columns)));
		let data = props.data;
		const { field, order, sorter } = this.state.sortData;

		if (!_.isNil(field) && order && _.isFunction(sorter)) {
			data = data.slice().sort(sorter);
			if (order === 'desc') {
				data.reverse();
			}
		}

		return (
			<RcTable
				{...props}
				data={data}
				prefixCls={props.prefixCls}
				columns={columns}
				rowClassName={this.getRowClassName}
			/>
		);
	}

	private get isAllSelected() {
		// FIXME: ts-ui 的逻辑和一般的情况相反，当有一行勾选后全选按钮就被勾选
		// const {
		//   rowSelection: { selectedRowKeys }
		// } = this.props;
		// const { data = [] } = this.props;
		// return (
		//   selectedRowKeys.length &&
		//   selectedRowKeys.length === data.length &&
		//   data.every(row => selectedRowKeys.includes(this.getRowKey(row)))
		// );

		return !!this.props.rowSelection!.selectedRowKeys.length;
	}

	private isSelected(row: T) {
		return this.props.rowSelection!.selectedRowKeys.includes(this.getRowKey(row));
	}

	private getColumnKey(column: Column<T>) {
		return (!_.isNil(column.key) ? column.key : column.dataIndex) as string | number;
	}

	private getRowKey(row: T) {
		if (_.isFunction(this.props.rowKey)) {
			return this.props.rowKey(row);
		}

		return _.get(row, this.props.rowKey);
	}

	private getRowClassName = (row: T, index: number, indent: string) => {
		const classNames = [];
		const { prefixCls, rowSelection, rowClassName } = this.props;

		// 添加上层的自定义 className
		if (rowClassName) {
			if (_.isString(rowClassName)) {
				classNames.push(rowClassName);
			} else if (_.isFunction(rowClassName)) {
				classNames.push(rowClassName(row, index, indent));
			}
		}

		// 添加勾选状态
		if (rowSelection && this.isSelected(row)) {
			classNames.push(`${prefixCls}-row-selected`);
		}

		return cx(classNames);
	};

	private renderSearch(columns: Array<Column<T>>) {
		const { prefixCls } = this.props;
		const { searchData, searchActiveColumn } = this.state;
		return columns.map(column => {
			if (!column.searchTriggers) {
				return column;
			}
			const key = this.getColumnKey(column);
			const patch = {
				title: (
					<SearchTitle
						column={column}
						prefixCls={prefixCls!}
						activeColumn={searchActiveColumn}
						setActiveColumn={this.setActiveColumn}
						columnKey={key}
						isActive={searchActiveColumn === key}
						value={searchData[key] || ''}
						onChange={(text: string) => {
							const data = { ...searchData };
							if (!text) {
								delete data[key];
							} else {
								data[key] = text;
							}
							this.setState({
								searchData: data
							});
						}}
						onSearch={() => {
							this.callChange();
						}}
					></SearchTitle>
				),
				className: cx(`${prefixCls}-search-cell`, column.className)
			};

			return _.merge({}, column, patch);
		});
	}

	private renderSort(columns: Array<Column<T>>) {
		const { prefixCls } = this.props;
		const { sortData } = this.state;
		return columns.map(column => {
			if (!column.sorter) {
				return column;
			}
			const key = this.getColumnKey(column);
			const directions: Order[] =
				column.sortDirections && column.sortDirections.length
					? column.sortDirections
					: Table.defaultSortDirections;
			let order = column.sortOrder;

			if (!order) {
				if (sortData.field === key) {
					order = sortData.order;
				} else {
					order = column.defaultSortOrder || false;
				}
			}
			const patch = {
				title: <SortTitle prefixCls={prefixCls!} column={column} order={order as Order | false} />,
				className: cx(
					{
						[`${prefixCls}-sort-cell--active`]: sortData.field === key && !!order
					},
					`${prefixCls}-sort-cell`,
					column.className
				),
				onHeaderCell: () => {
					let originCellProps: any = {};
					if (column.onHeaderCell) {
						originCellProps = column.onHeaderCell(column);
					}
					return {
						...originCellProps,
						onClick: (e: Event) => {
							if (originCellProps.onClick) {
								originCellProps.onClick(e);
							}

							const orderIndex = directions.indexOf(order as Order); // if order = false, index = -1
							let nextOrder: Order | false = false;
							if (orderIndex < directions.length - 1) {
								nextOrder = directions[orderIndex + 1];
							}
							this.setState({
								sortData: {
									field: key,
									order: nextOrder,
									sorter: column.sorter
								}
							});
							this.callChange();
						}
					};
				}
			};

			return _.merge({}, column, patch);
		});
	}

	private renderRowSelections(columns: Array<Column<T>>) {
		const { rowSelection, prefixCls } = this.props;
		const columnsClone = columns.slice();
		if (rowSelection) {
			columnsClone.unshift({
				key: 'selection-column',
				className: prefixCls + '-selection-column',
				width: 40,
				title: <i className={checkboxClassName(prefixCls!, this.isAllSelected)} />,
				onCell: this.onSelectionCell,
				onHeaderCell: this.onSelectionHeaderCell,
				render: this.renderSelectionBox(this.props.rowSelection!.type || 'checkbox')
			});

			return columnsClone;
		}

		return columns;
	}

	private renderSelectionBox(type: RowSelection<T>['type']) {
		return (value: any, row: T, index: number) => {
			switch (type) {
				case 'checkbox':
					return <i className={checkboxClassName(this.props.prefixCls!, this.isSelected(row))} />;
				default:
					return undefined;
			}
		};
	}

	private onSelectionHeaderCell = () => {
		return {
			onClick: () => {
				this.toggleSelectAllRows();
			}
		};
	};

	private onSelectionCell = (record: T, rowIndex: number) => {
		return {
			onClick: () => {
				const row = this.props.data[rowIndex];
				this.toggleSelectRow(row);
			}
		};
	};

	private toggleSelectAllRows() {
		if (!this.props.data || this.isAllSelected) {
			this.updateSelectedRows([]);
		} else {
			this.updateSelectedRows(this.props.data);
		}
	}

	private toggleSelectRow(row: T) {
		const key = this.getRowKey(row);
		const selectedRowKeys = this.props.rowSelection!.selectedRowKeys.slice();
		const index = selectedRowKeys.indexOf(key);

		if (index > -1) {
			selectedRowKeys.splice(index, 1);
		} else {
			selectedRowKeys.push(key);
		}

		const rows: T[] = selectedRowKeys.map(
			key => this.props.data.find(row => key === this.getRowKey(row)) as T
		);

		this.updateSelectedRows(rows);
	}

	private updateSelectedRows(rows: T[]) {
		if (!Array.isArray(rows)) {
			return;
		}

		const { rowSelection } = this.props;
		const keys = rows.map(this.getRowKey, this);
		rowSelection!.onChange(keys, rows);
	}

	private setActiveColumn = (key?: string | number) => {
		this.setState({
			searchActiveColumn: key
		});
	};

	private callChange() {
		Promise.resolve().then(() => {
			if (this.props.onChange) {
				this.props.onChange({
					search: this.state.searchData,
					sort: this.state.sortData
				});
			}
		});
	}
}

export default Table;
