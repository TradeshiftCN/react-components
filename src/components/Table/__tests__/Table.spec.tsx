import React from 'react';
import { shallow, mount } from 'enzyme';
import flushPromises from 'flush-promises';
import Table from '..';

describe('Table', () => {
	it('basic', () => {
		const wrapper = shallow(
			<Table
				data={[{ id: 1, value: 11 }, { id: 2, value: 22 }]}
				columns={[{ title: 'id', dataIndex: 'id' }, { title: 'value', dataIndex: 'value' }]}
				rowKey="id"
			/>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it('selection', async () => {
		const onChange = jest.fn((keys: number[]) => {
			wrapper.setProps({ ...wrapper.props(), rowSelection: { onChange, selectedRowKeys: keys } });
		});
		const wrapper = mount(
			<Table
				data={[{ id: 1, value: 11 }, { id: 2, value: 22 }]}
				columns={[{ title: 'id', dataIndex: 'id' }, { title: 'value', dataIndex: 'value' }]}
				rowKey="id"
				rowSelection={{ onChange, selectedRowKeys: [] }}
			/>
		);

		// select first row
		wrapper
			.find('tbody tr')
			.first()
			.find('.ts-rc-table-selection-icon')
			.simulate('click');
		await flushPromises();
		expect(onChange).toHaveBeenCalledTimes(1);
		expect(onChange.mock.calls[0]).toEqual([[1], [{ id: 1, value: 11 }]]);
		expect(wrapper.find('thead .ts-rc-table-selection-icon--checked').exists()).toBe(true);

		// deselect all rows
		wrapper.find('thead .ts-rc-table-selection-icon').simulate('click');
		await flushPromises();
		expect(onChange).toHaveBeenCalledTimes(2);
		expect(onChange.mock.calls[1]).toEqual([[], []]);
		expect(wrapper.find('thead .ts-rc-table-selection-icon--checked').exists()).toBe(false);

		// select all rows
		wrapper.find('thead .ts-rc-table-selection-icon').simulate('click');
		await flushPromises();
		expect(onChange).toHaveBeenCalledTimes(3);
		expect(onChange.mock.calls[2]).toEqual([[1, 2], [{ id: 1, value: 11 }, { id: 2, value: 22 }]]);
		expect(wrapper.find('thead .ts-rc-table-selection-icon--checked').exists()).toBe(true);
	});

	it('search', async () => {
		const onChange = jest.fn();
		const wrapper = mount(
			<Table
				data={[{ id: 1, value: 11 }, { id: 2, value: 22 }]}
				columns={[
					{ title: 'id', dataIndex: 'id', searchTriggers: true },
					{ title: 'value', dataIndex: 'value', searchTriggers: ['input'] }
				]}
				rowKey="id"
				onChange={onChange}
			/>
		);

		expect(wrapper.find('thead .ts-icon-search').length).toBe(2);

		wrapper
			.find('thead .ts-icon-search')
			.first()
			.simulate('click');
		await flushPromises();
		expect(
			wrapper
				.find('thead th')
				.first()
				.find('input')
				.exists()
		).toBe(true);

		wrapper
			.find('thead th')
			.first()
			.find('input')
			.simulate('change', { target: { value: 'keyword' } })
			.simulate('keydown', { keyCode: 13 });
		await flushPromises();
		expect(onChange).toHaveBeenCalledTimes(1);
		expect(onChange.mock.calls[0][0].search).toEqual({ id: 'keyword' });
	});

	it('sort', async () => {
		const wrapper = mount(
			<Table
				data={[{ id: 1, value: 11 }, { id: 2, value: 22 }]}
				columns={[
					{ title: 'id', dataIndex: 'id', sorter: true },
					{ title: 'value', dataIndex: 'value', sorter: (a, b) => a.id - b.id }
				]}
				rowKey="id"
			/>
		);

		wrapper
			.find('thead .ts-rc-table-sort-cell')
			.first()
			.simulate('click');
		await flushPromises();
		expect(wrapper.find('.ts-rc-table-sort-title-icon--asc').exists()).toBe(true);

		wrapper
			.find('thead .ts-rc-table-sort-cell')
			.first()
			.simulate('click');
		expect(wrapper.find('.ts-rc-table-sort-title-icon--desc').exists()).toBe(true);

		wrapper
			.find('thead .ts-rc-table-sort-cell')
			.first()
			.simulate('click');
		expect(wrapper.find('.ts-rc-table-sort-title-icon--asc').exists()).toBe(false);
		expect(wrapper.find('.ts-rc-table-sort-title-icon--desc').exists()).toBe(false);
	});
});
