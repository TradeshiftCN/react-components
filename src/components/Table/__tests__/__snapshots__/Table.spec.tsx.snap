// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table basic 1`] = `
<Table
  bodyStyle={Object {}}
  columns={
    Array [
      Object {
        "dataIndex": "id",
        "title": "id",
      },
      Object {
        "dataIndex": "value",
        "title": "value",
      },
    ]
  }
  data={
    Array [
      Object {
        "id": 1,
        "value": 11,
      },
      Object {
        "id": 2,
        "value": 22,
      },
    ]
  }
  emptyText=""
  expandIcon={[Function]}
  expandIconAsCell={false}
  expandIconColumnIndex={0}
  onHeaderRow={[Function]}
  onRow={[Function]}
  prefixCls="ts-rc-table"
  rowClassName={[Function]}
  rowKey="id"
  rowRef={[Function]}
  scroll={Object {}}
  showHeader={true}
  style={Object {}}
  useFixedHeader={false}
/>
`;
